vllm/distributed/device_communicators/quick_all_reduce.py

```diff
@@ -0,0 +1,122 @@
+# SPDX-License-Identifier: Apache-2.0
+import logging
+import os
+from typing import Union
+
+import torch
+import torch.distributed as dist
+from torch.distributed import ProcessGroup
+
+from vllm import _custom_ops as ops
+from vllm.platforms import current_platform
+
+logger = logging.getLogger(__name__)
+
+try:
+    ops.qr_max_size()
+    ops_available = True
+except Exception:
+    # For CPUs
+    ops_available = False
+
+
+class QuickAllReduce:
+    _SUPPORTED_WORLD_SIZES = [2, 4, 8]
+    _SUPPORTED_DTYPES = [torch.float16, torch.bfloat16]
+
+    def __init__(self, group: ProcessGroup,
+                 device: Union[int, str, torch.device]) -> None:
+        self.disabled = True
+        if not ops_available:
+            # disable because of missing quick reduce library
+            # e.g. in a non-cuda environment
+            logger.info("Custom allreduce is disabled because "
+                        "of missing custom allreduce library")
+            return
+
+        self.max_size = ops.qr_max_size()
+        self.group = group
+        self.quantized = os.environ.get("VLLM_ROCM_CA_QUANTIZED", "0") == "1"
+
+        # On RocM bfloat16 kernels are slower than fp16
+        # due to slower match operations
+        # If environment is not set to 1 we convert input to fp16
+        self.use_bf16_kernels = os.environ.get("VLLM_ROCM_CA_BF16_KERNELS",
+                                               "0") == "1"
```

Comment:
This should also be moved to `envs.py`. How much slower is `fp16` than `bf16`? Could you add a comment so that people know what to expect?